server/src/main/java/org/elasticsearch/cluster/ClusterInfoSimulator.java

```diff
@@ -35,13 +35,48 @@ public class ClusterInfoSimulator {
 
     public ClusterInfoSimulator(RoutingAllocation allocation) {
         this.allocation = allocation;
-        this.leastAvailableSpaceUsage = new HashMap<>(allocation.clusterInfo().getNodeLeastAvailableDiskUsages());
-        this.mostAvailableSpaceUsage = new HashMap<>(allocation.clusterInfo().getNodeMostAvailableDiskUsages());
+        this.leastAvailableSpaceUsage = getAdjustedDiskSpace(allocation, allocation.clusterInfo().getNodeLeastAvailableDiskUsages());
+        this.mostAvailableSpaceUsage = getAdjustedDiskSpace(allocation, allocation.clusterInfo().getNodeMostAvailableDiskUsages());
         this.shardSizes = new CopyOnFirstWriteMap<>(allocation.clusterInfo().shardSizes);
         this.shardDataSetSizes = Map.copyOf(allocation.clusterInfo().shardDataSetSizes);
         this.dataPath = Map.copyOf(allocation.clusterInfo().dataPath);
     }
 
+    /**
+     * Cluster info contains a reserved space that is necessary to finish initializing shards (that are currently in progress).
+     * for all initializing shards sum(expected size) = reserved space + already used space
+     * This deducts already used space from disk usage as when shard start is simulated it is going to add entire expected shard size.
+     */
+    private static Map<String, DiskUsage> getAdjustedDiskSpace(RoutingAllocation allocation, Map<String, DiskUsage> diskUsage) {
+        var diskUsageCopy = new HashMap<>(diskUsage);
+        for (var entry : diskUsageCopy.entrySet()) {
+            var nodeId = entry.getKey();
+            var usage = entry.getValue();
+
+            var reserved = allocation.clusterInfo().getReservedSpace(nodeId, usage.path());
+            if (reserved.getTotal() == 0) {
+                continue;
+            }
```

Comment:
I am surprised reserved could include shards with 0 remaining sizes.

I wonder if it makes sense to update below:

https://github.com/elastic/elasticsearch/blob/997fbb372cdec82eb803d32de5147dbccec21e64/server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java#L468-L474

to only include sizes strictly greater then 0?